src/hotspot/share/gc/z/zRelocationSet.cpp

```diff
@@ -22,35 +22,50 @@
  */
 
 #include "precompiled.hpp"
-#include "gc/z/zForwarding.hpp"
+#include "gc/z/zForwarding.inline.hpp"
+#include "gc/z/zForwardingAllocator.inline.hpp"
 #include "gc/z/zRelocationSet.hpp"
+#include "gc/z/zStat.hpp"
 #include "memory/allocation.hpp"
 
 ZRelocationSet::ZRelocationSet() :
+    _allocator(),
     _forwardings(NULL),
     _nforwardings(0) {}
 
-void ZRelocationSet::populate(ZPage* const* group0, size_t ngroup0,
-                              ZPage* const* group1, size_t ngroup1) {
-  _nforwardings = ngroup0 + ngroup1;
-  _forwardings = REALLOC_C_HEAP_ARRAY(ZForwarding*, _forwardings, _nforwardings, mtGC);
+void ZRelocationSet::populate(ZPage* const* small, size_t nsmall,
+                              ZPage* const* medium, size_t nmedium,
+                              size_t forwarding_entries) {
+  // Set relocation set length
+  _nforwardings = nsmall + nmedium;
 
+  // Initialize forwarding allocator to have room for the
+  // relocation set, all forwardings, and all forwarding entries.
+  const size_t relocation_set_size = _nforwardings * sizeof(ZForwarding*);
+  const size_t forwardings_size = _nforwardings * sizeof(ZForwarding);
+  const size_t forwarding_entries_size = forwarding_entries * sizeof(ZForwardingEntry);
+  _allocator.reset(relocation_set_size + forwardings_size + forwarding_entries_size);
+
+  // Allocate relocation set
+  _forwardings = new (_allocator.alloc(relocation_set_size)) ZForwarding*[_nforwardings];
+
+  // Populate relocation set array
   size_t j = 0;
 
-  // Populate group 0
-  for (size_t i = 0; i < ngroup0; i++) {
-    _forwardings[j++] = ZForwarding::create(group0[i]);
+  // Populate medium pages
+  for (size_t i = 0; i < nmedium; i++) {
+    _forwardings[j++] = ZForwarding::alloc(&_allocator, medium[i]);
   }
 
-  // Populate group 1
-  for (size_t i = 0; i < ngroup1; i++) {
-    _forwardings[j++] = ZForwarding::create(group1[i]);
+  // Populate small pages
+  for (size_t i = 0; i < nsmall; i++) {
+    _forwardings[j++] = ZForwarding::alloc(&_allocator, small[i]);
   }
+
+  // Update statistics
+  ZStatRelocation::set_at_populate_relocation_set(_allocator.size());
```

Comment:
Instead of `_allocator.size()`, one could use the same arg of `_allocator.reset()`, which is already known in this scope.

Reply:
Sure, I can change that.

Reply:
Actually, the `size()` function will still be needed in the follow up patch #805, where the size isn't calculated in the same score. So in the end I think it's easier to just keep it here too.

Reply:
OK.